// @cycle/rx-adapter/dist/cycle-rx-adapter.js
(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
(function (global){
"use strict";

var Rx = typeof window !== "undefined" ? window['Rx'] : typeof global !== "undefined" ? global['Rx'] : null;
var RxJSAdapter = {
    adapt: function adapt(originStream, originStreamSubscribe) {
        if (this.isValidStream(originStream)) {
            return originStream;
        }
        return Rx.Observable.create(function (destinationObserver) {
            var originObserver = {
                next: function next(x) {
                    return destinationObserver.onNext(x);
                },
                error: function error(e) {
                    return destinationObserver.onError(e);
                },
                complete: function complete() {
                    return destinationObserver.onCompleted();
                }
            };
            var dispose = originStreamSubscribe(originStream, originObserver);
            return function () {
                if (typeof dispose === 'function') {
                    dispose.call(null);
                }
            };
        });
    },
    remember: function remember(observable) {
        return observable.shareReplay(1);
    },
    makeSubject: function makeSubject() {
        var stream = new Rx.Subject();
        var observer = {
            next: function next(x) {
                stream.onNext(x);
            },
            error: function error(err) {
                stream.onError(err);
            },
            complete: function complete(x) {
                stream.onCompleted();
            }
        };
        return { stream: stream, observer: observer };
    },
    isValidStream: function isValidStream(stream) {
        return typeof stream.subscribeOnNext === 'function' && typeof stream.onValue !== 'function';
    },
    streamSubscribe: function streamSubscribe(stream, observer) {
        var subscription = stream.subscribe(function (x) {
            return observer.next(x);
        }, function (e) {
            return observer.error(e);
        }, function (x) {
            return observer.complete(x);
        });
        return function () {
            subscription.dispose();
        };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = RxJSAdapter;


}).call(this,typeof global !== "undefined" ? global : typeof self !== "undefined" ? self : typeof window !== "undefined" ? window : {})
},{}]},{},[1]);
