// @cycle/dom/lib/es6/utils.js
function isValidNode(obj) {
    var ELEM_TYPE = 1;
    var FRAG_TYPE = 11;
    return typeof HTMLElement === 'object'
        ? obj instanceof HTMLElement || obj instanceof DocumentFragment
        : obj &&
            typeof obj === 'object' &&
            obj !== null &&
            (obj.nodeType === ELEM_TYPE || obj.nodeType === FRAG_TYPE) &&
            typeof obj.nodeName === 'string';
}
export function isClassOrId(str) {
    return str.length > 1 && (str[0] === '.' || str[0] === '#');
}
export function isDocFrag(el) {
    return el.nodeType === 11;
}
export var SCOPE_PREFIX = '$$CYCLEDOM$$-';
export function checkValidContainer(container) {
    if (typeof container !== 'string' && !isValidNode(container)) {
        throw new Error('Given container is not a DOM element neither a selector string.');
    }
}
export function getValidNode(selectors) {
    var domElement = typeof selectors === 'string'
        ? document.querySelector(selectors)
        : selectors;
    if (typeof selectors === 'string' && domElement === null) {
        throw new Error("Cannot render into unknown element `" + selectors + "`");
    }
    return domElement;
}
/**
 * The full scope of a namespace is the "absolute path" of scopes from
 * parent to child. This is extracted from the namespace, filter only for
 * scopes in the namespace.
 */
export function getFullScope(namespace) {
    return namespace
        .filter(function (c) { return c.indexOf(SCOPE_PREFIX) > -1; })
        .map(function (c) { return c.replace(SCOPE_PREFIX, ''); })
        .join('-');
}
export function getSelectors(namespace) {
    return namespace.filter(function (c) { return c.indexOf(SCOPE_PREFIX) === -1; }).join(' ');
}
